### PR TITLE
make metric.Connection thread safe

### PR DESCRIPTION
**Pull request by m25n**

### CODE REVIEW

File: internal/wavefront/metric/connection.go [hunk 1]
```diff
@@ -64,35 +66,62 @@ func (c *Connection) Connect(addr string) error {
...
+}
+
+// close is not thread safe and must only be called when already holding c.mu
+func (c *Connection) close() {
+	if c.connected() {
+		c.sender.Close()
 	}
-	c.sender.Close()
 	c.sender = nil
 	c.addr = ""
```
Comment (johncornish) [inline internal/wavefront/metric/connection.go:126]: So it's okay to run these outside of the c.connected() check?

File: internal/wavefront/metric/connection_test.go [hunk 2]
```diff
@@ -167,10 +168,38 @@ func TestConnection(t *testing.T) {
...
-		conn.FlushMetrics()
+		conn.Flush()
 
 		require.Equal(t, 1, len(mockSender.SentMetrics))
 	})
+
+	t.Run("handles concurrency", func(t *testing.T) {
+		conn := NewTestConnection(&testhelper.StubSender{})
+
+		const runs = 100_000
```
Comment (johncornish) [inline internal/wavefront/metric/connection_test.go:179]: 😯👍

**Review by johncornish — APPROVED** [review]

File: internal/wavefront/metric/connection.go (hunk 1 above)
Comment (johncornish) [reply internal/wavefront/metric/connection.go:126]: sender = nil and addr = "" that is

Comment (m25n) [reply internal/wavefront/metric/connection.go:126]: I think so. If sender or addr are already empty, then making them empty again should be benign.

File: internal/wavefront/metric/connection_test.go (hunk 2 above)
Comment (m25n) [reply internal/wavefront/metric/connection_test.go:179]: Yeah, this is a probabilistic test. However, 100k runs seems to drive the probability of triggering a concurrency bug to something pretty close to 100%. This test runs in like 60ms which seems acceptable.

File: internal/wavefront/metric/connection.go (hunk 1 above)
Comment (m25n) [reply internal/wavefront/metric/connection.go:124]: Your comment made me realize that I didn't need the `if c.connected()` check in the `Connect` method. I went ahead and removed it.

**Review by mmichael — COMMENTED** [review]
Looks good to me

**Pull request merged**